### PR TITLE
tree: if matcher says that tree should be skipped, skip it

### PR DESCRIPTION
**Pull request by martinvonz**
When using a sparse working copy (e.g. with no files at all) and updating the working copy from the root commit to a commit with millions of files, we shouldn't have to walk the parts of the diff that doesn't match the sparse patterns. However, we still do the full walk because our `Tree::diff()` currently doesn't care about what the matcher tells us to visit, it only filters out unwanted files after visiting them. This commit fixes that for the special (but common) case of matching nothing in a directory.

I tried also adding special handling for when the matcher says that we should only visit a few entries, but it wasn't clearly better in the cases I tested it on. I'll keep that patch around and might send it if I find some cases where it helps.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/tree.rs [hunk 1]
```diff
@@ -309,87 +309,99 @@ impl<T> Diff<T> {
...
-                                    Some(before.value()),
-                                    Some(after.value()),
-                                ));
-                            }
+                        (None, None) => {
+                            // both iterators exhausted
+                            return None;
                         }
                     }
                 }
```
Comment (martinvonz) [inline lib/src/tree.rs:394]: It's hard to tell from the diff, but the only change inside the loop is that `self.it` became `it`

File: lib/src/tree.rs [hunk 2]
```diff
@@ -309,87 +309,99 @@ impl<T> Diff<T> {
...
-            _matcher: matcher,
+impl<'trees> TreeEntryDiffIterator<'trees> {
+    fn new(tree1: &'trees Tree, tree2: &'trees Tree, matcher: &dyn Matcher) -> Self {
+        match matcher.visit(tree1.dir()) {
+            Visit::Nothing => TreeEntryDiffIterator::NoWalk,
+            _ => {
+                let it1 = tree1.entries_non_recursive().peekable();
+                let it2 = tree2.entries_non_recursive().peekable();
+                TreeEntryDiffIterator::WalkEntries { it1, it2 }
+            }
```
Comment (yuja) [inline lib/src/tree.rs:329]: I don't know how these enum variants will expand, but at this point, I think it's better to leave `TreeEntryDiffIterator::WalkEntries` as a struct (without a matcher), and let the caller decide whether or not to visit the tree pair.

**Review by yuja — APPROVED** [review]

File: lib/src/tree.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/tree.rs]: I did that as a separate commit. See if you prefer the result and I'll squash them into one. I'm fine with either.

Comment (yuja) [reply lib/src/tree.rs]: Thanks. I prefer the new version as it clarifies `(Maybe)TreeEntryDiffIterator` variants never change during iteration. 
We can also use `Option<TreeEntryDiffIterator>`, but I'm not sure if that's better.

Comment (martinvonz) [reply lib/src/tree.rs]: I've squashed the two commits. Thanks for reviewing.

**Pull request merged**